File: mobile/lib/core/utils/error_handler.dart
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

import '../errors/api_exceptions.dart';
import '../network/network_utils.dart';

/// Centralized error handling utility
class ErrorHandler {
  /// Get a user-friendly error message from an exception
  static String getUserFriendlyMessage(Exception error) {
    if (error is UnauthorizedException) {
      return 'Your session has expired. Please log in again.';
    } else if (error is ForbiddenException) {
      return 'You don\'t have permission to access this resource.';
    } else if (error is NotFoundException) {
      return 'The requested resource was not found.';
    } else if (error is BadRequestException) {
      return 'Invalid request. Please check your input and try again.';
    } else if (error is ServerException) {
      return 'Server error. Please try again later.';
    } else if (error is TimeoutException) {
      return 'Connection timed out. Please check your internet connection and try again.';
    } else if (error is NetworkException) {
      return 'No internet connection. Please check your network settings and try again.';
    } else if (error is CircuitBreakerException) {
      return 'Too many failed requests. Please try again later.';
    } else if (error is RequestCancelledException) {
      return 'Request was cancelled.';
    } else if (error is ApiException) {
      return error.message;
    } else {
      return 'An unexpected error occurred. Please try again.';
    }
  }

  /// Get a detailed error message for logging
  static String getDetailedErrorMessage(dynamic error) {
    if (error is DioException) {
      return _handleDioError(error);
    } else if (error is SocketException) {
      return 'SocketException: ${error.message}';
    } else if (error is Exception) {
      return 'Exception: ${error.toString()}';
    } else {
      return 'Unknown error: ${error.toString()}';
    }
  }

  /// Handle Dio errors and return detailed message
  static String _handleDioError(DioException error) {
    final buffer = StringBuffer('DioException: ');
    buffer.write('Type: ${error.type}, ');
    buffer.write('Message: ${error.message}, ');
    
    if (error.response != null) {
      buffer.write('Status: ${error.response?.statusCode}, ');
      
      if (error.response?.data != null) {
        if (error.response?.data is Map) {
          final data = error.response?.data as Map;
          if (data.containsKey('detail')) {
            buffer.write('Detail: ${data['detail']}');
          } else if (data.containsKey('message')) {
            buffer.write('Detail: ${data['message']}');
          } else {
            buffer.write('Data: ${error.response?.data}');
          }
        } else {
          buffer.write('Data: ${error.response?.data}');
        }
      }
    }
    
    return buffer.toString();
  }

  /// Show a snackbar with an error message
  static void showErrorSnackBar(BuildContext context, String message) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text(message),
        backgroundColor: Colors.red,
        behavior: SnackBarBehavior.floating,
        duration: const Duration(seconds: 4),
        action: SnackBarAction(
          label: 'Dismiss',
          textColor: Colors.white,
          onPressed: () {
            ScaffoldMessenger.of(context).hideCurrentSnackBar();
          },
        ),
      ),
    );
  }

  /// Show a dialog with an error message
  static Future<void> showErrorDialog(
    BuildContext context, {
    required String title,
    required String message,
    String? buttonText,
  }) async {
    return showDialog<void>(
      context: context,
      barrierDismissible: false,
      builder: (BuildContext context) {
        return AlertDialog(
          title: Text(title),
          content: SingleChildScrollView(
            child: Text(message),
          ),
          actions: <Widget>[
            TextButton(
              child: Text(buttonText ?? 'OK'),
              onPressed: () {
                Navigator.of(context).pop();
              },
            ),
          ],
        );
      },
    );
  }

  /// Log an error with detailed information
  static void logError(dynamic error, [StackTrace? stackTrace]) {
    if (kDebugMode) {
      print('ERROR: ${getDetailedErrorMessage(error)}');
      if (stackTrace != null) {
        print('STACK TRACE: $stackTrace');
      }
    }
    
    // Here you could add integration with a logging service like Firebase Crashlytics
    // or Sentry for production error tracking
  }

  /// Handle an error with appropriate UI feedback
  static void handleError(
    BuildContext context,
    dynamic error, {
    String? fallbackMessage,
    bool showSnackBar = true,
    bool showDialog = false,
    String? dialogTitle,
    VoidCallback? onRetry,
  }) {
    // Log the error
    logError(error);
    
    // Get user-friendly message
    String message;
    if (error is Exception) {
      message = getUserFriendlyMessage(error);
    } else {
      message = fallbackMessage ?? 'An unexpected error occurred. Please try again.';
    }
    
    // Show UI feedback
    if (showSnackBar) {
      showErrorSnackBar(context, message);
    }
    
    if (showDialog) {
      showErrorDialog(
        context,
        title: dialogTitle ?? 'Error',
        message: message,
        buttonText: onRetry != null ? 'Retry' : 'OK',
      ).then((_) {
        if (onRetry != null) {
          onRetry();
        }
      });
    }
  }
}

File: web/src/utils/errorHandler.js
/**
 * Centralized error handling utility
 * Provides consistent error handling across the application
 */

// Error types
export const ErrorTypes = {
  NETWORK: 'NETWORK',
  AUTH: 'AUTH',
  VALIDATION: 'VALIDATION',
  SERVER: 'SERVER',
  TIMEOUT: 'TIMEOUT',
  UNKNOWN: 'UNKNOWN',
};

/**
 * Get error type from an error object
 * @param {Error} error - Error object
 * @returns {string} Error type
 */
export const getErrorType = (error) => {
  if (!error) return ErrorTypes.UNKNOWN;
  
  // Network errors
  if (error.message === 'Network Error' || 
      error.code === 'ERR_NETWORK' ||
      error.name === 'NetworkError') {
    return ErrorTypes.NETWORK;
  }
  
  // Timeout errors
  if (error.code === 'ECONNABORTED' || 
      error.message?.includes('timeout')) {
    return ErrorTypes.TIMEOUT;
  }
  
  // Authentication errors
  if (error.response?.status === 401 || 
      error.response?.status === 403 ||
      error.message?.includes('unauthorized') ||
      error.message?.includes('forbidden')) {
    return ErrorTypes.AUTH;
  }
  
  // Validation errors
  if (error.response?.status === 400 || 
      error.response?.status === 422) {
    return ErrorTypes.VALIDATION;
  }
  
  // Server errors
  if (error.response?.status >= 500) {
    return ErrorTypes.SERVER;
  }
  
  return ErrorTypes.UNKNOWN;
};

/**
 * Get user-friendly error message
 * @param {Error} error - Error object
 * @returns {string} User-friendly error message
 */
export const getUserFriendlyMessage = (error) => {
  if (!error) return 'An unknown error occurred';
  
  // If the error already has a user-friendly message, use it
  if (error.userMessage) {
    return error.userMessage;
  }
  
  const errorType = getErrorType(error);
  
  switch (errorType) {
    case ErrorTypes.NETWORK:
      return 'Network error. Please check your internet connection and try again.';
    
    case ErrorTypes.TIMEOUT:
      return 'Request timed out. The server might be down or overloaded. Please try again later.';
    
    case ErrorTypes.AUTH:
      if (error.response?.status === 401) {
        return 'Your session has expired. Please log in again.';
      }
      if (error.response?.status === 403) {
        return 'You don\'t have permission to access this resource.';
      }
      return 'Authentication error. Please log in again.';
    
    case ErrorTypes.VALIDATION:
      // Try to extract validation error messages
      if (error.response?.data?.detail) {
        return error.response.data.detail;
      }
      if (error.response?.data?.message) {
        return error.response.data.message;
      }
      return 'Invalid input. Please check your information and try again.';
    
    case ErrorTypes.SERVER:
      return 'Server error. Our team has been notified. Please try again later.';
    
    default:
      // Try to extract any message from the error
      if (error.response?.data?.detail) {
        return error.response.data.detail;
      }
      if (error.response?.data?.message) {
        return error.response.data.message;
      }
      if (error.message && !error.message.includes('Error') && !error.message.includes('error')) {
        return error.message;
      }
      return 'An unexpected error occurred. Please try again.';
  }
};

/**
 * Get detailed error message for logging
 * @param {Error} error - Error object
 * @returns {string} Detailed error message
 */
export const getDetailedErrorMessage = (error) => {
  if (!error) return 'Unknown error';
  
  const details = [];
  
  // Add error message
  if (error.message) {
    details.push(`Message: ${error.message}`);
  }
  
  // Add error name and code
  if (error.name) details.push(`Name: ${error.name}`);
  if (error.code) details.push(`Code: ${error.code}`);
  
  // Add response details if available
  if (error.response) {
    details.push(`Status: ${error.response.status}`);
    
    if (error.response.data) {
      if (typeof error.response.data === 'object') {
        details.push(`Data: ${JSON.stringify(error.response.data)}`);
      } else {
        details.push(`Data: ${error.response.data}`);
      }
    }
  }
  
  // Add request details if available
  if (error.config) {
    details.push(`URL: ${error.config.url}`);
    details.push(`Method: ${error.config.method}`);
  }
  
  return details.join(', ');
};

/**
 * Log an error with detailed information
 * @param {Error} error - Error object
 * @param {string} context - Context where the error occurred
 */
export const logError = (error, context = '') => {
  const errorType = getErrorType(error);
  const detailedMessage = getDetailedErrorMessage(error);
  
  console.error(`[${context}] ${errorType} Error:`, detailedMessage);
  
  // Here you could add integration with a logging service like Sentry
  // for production error tracking
};

/**
 * Create an enhanced error with user-friendly message
 * @param {Error} error - Original error
 * @param {string} userMessage - Optional custom user message
 * @returns {Error} Enhanced error
 */
export const createEnhancedError = (error, userMessage = null) => {
  const enhancedError = new Error(userMessage || getUserFriendlyMessage(error));
  enhancedError.originalError = error;
  enhancedError.errorType = getErrorType(error);
  enhancedError.userMessage = userMessage || getUserFriendlyMessage(error);
  enhancedError.timestamp = new Date().toISOString();
  return enhancedError;
};

/**
 * Handle an error with appropriate logging and return enhanced error
 * @param {Error} error - Error object
 * @param {string} context - Context where the error occurred
 * @param {string} userMessage - Optional custom user message
 * @returns {Error} Enhanced error
 */
export const handleError = (error, context = '', userMessage = null) => {
  // Log the error
  logError(error, context);
  
  // Create and return enhanced error
  return createEnhancedError(error, userMessage);
};
